Add tests based on einops paper

RomeoV opened the pull request:
The einops paper published at ICLR 2022 presents a list of examples comparing numpy and einops.

Here we implement all those exmaples in TensorCast. The last tests involving repeat patterns were enabled by 6f23e318ba668a6fe76b1ffa1a4f3c83e4fc7ea6.

See also #53.

RomeoV committed:
Add tests based on einops paper

The einops paper published at ICLR 2022 presents a list of examples
comparing numpy and einops.

Here we implement all those exmaples in TensorCast.
The last tests involving repeat patterns were enabled by 6f23e318ba668a6fe76b1ffa1a4f3c83e4fc7ea6.

mcabbott commented on test/einops_comparison.jl:
```diff
@@ -0,0 +1,182 @@
+  @testset "Test 11" begin
+    x = rand(2, 2, 3, 4)
+    @test (
+      maximum(x, dims=[2, 3])[:, 1, 1, :] == 
+      @reduce _[b, c] := maximum(h, w) x[b, h, w, c]
```
```suggestion
      maximum(x, dims=[2, 3])[:, 1, 1, :] == 
      dropdims(maximum(x, dims=(2, 3)), dims=(2, 3)) == 
      @reduce _[b, c] := maximum(h, w) x[b, h, w, c]
```

RomeoV replied:
Done in 2946fca2618522d4b84cc5a8231b8cab69ed92c5.
I like showcasing (here and later) how TensorCast.jl can reduce multiple ways of writing Julia code.
However, on a more general level, we should also remind ourselves that introducing more ways to do the same thing is not always good for a language.

mcabbott commented on test/einops_comparison.jl:
```diff
@@ -0,0 +1,182 @@
+  @testset "Test 7" begin
+    x = rand(3, 4, 5)
+    @test (
+      x[:] == 
+      @cast _[(b, t, c)] := x[b, t, c]
```
```suggestion
      x[:] == vec(x) == 
      @cast _[(b, t, c)] := x[b, t, c]
```

RomeoV replied:
Done in 2946fca2618522d4b84cc5a8231b8cab69ed92c5.

mcabbott commented on test/einops_comparison.jl:
```diff
@@ -0,0 +1,182 @@
+  @testset "Test 5" begin
+    xs = [rand(2, 2) for _ in 1:3]
+    @test (
+      cat(xs..., dims=3) ==
```
It's possible that we can have this
```suggestion
      cat(xs..., dims=3) ==
      stack(xs) ==
```
by adding using Compat somewhere

RomeoV replied:
Done in 2946fca2618522d4b84cc5a8231b8cab69ed92c5.

mcabbott commented on test/einops_comparison.jl:
```diff
@@ -0,0 +1,182 @@
+  @testset "Test 3" begin
+    x = rand(1, 2, 3, 4)
+    @test (
+      reshape(x, 2, 3, 4) ==
```
```suggestion
      reshape(x, 2, 3, 4) ==
      dropdims(x, dims=1) ==
```

RomeoV replied:
Done in 2946fca2618522d4b84cc5a8231b8cab69ed92c5.

mcabbott commented on test/einops_comparison.jl:
```diff
@@ -0,0 +1,182 @@
+published at ICLR 2022'; Listing 1.
+https://openreview.net/pdf?id=oapKSVM2bcj
+"""
+
```
Maybe it's nice to make this file self-contained, for anyone reading it as documentation. Also, loading Compat to use `stack` below:
```suggestion

using TensorCast, Test, Statistics, Compat

```

RomeoV replied:
Generally I agree, but don't super like the global imports. I tried to be more clear in 204b0632e6276dad4ef6d2cd937b1c6909fa7cf7.
Feel free to comment on this, as my approach seems somewhat uncommon in the Julia community.

mcabbott replied:
That's great, even clearer.

mcabbott commented on test/einops_comparison.jl:
```diff
@@ -0,0 +1,182 @@
+    @test (
+      [mean(x)] == 
+      @reduce _[_] := mean(b, h, w, c) x[b, h, w, c]
+    )
```
I'm not entirely sure what the python code returns here, but perhaps we should also show complete reduction to a number:
```suggestion
    @test (
      [mean(x)] == 
      @reduce _[_] := mean(b, h, w, c) x[b, h, w, c]  # reduce to a 1-element vector
    )
    @test (
      mean(x) ==
      @reduce _ := mean(b, h, w, c) x[b, h, w, c]  # reduce to a number
    )
```

RomeoV replied:
Done in 2946fca2618522d4b84cc5a8231b8cab69ed92c5.
I didn't realize this worked, this is quite nice!.

mcabbott reviewed (commented):
This is really nice, thanks!

I made a few suggestions mostly where I think showing an alternative might be helpful.

RomeoV committed:
Add mcabbott's suggestions

1) Showcase how `TensorCast.jl` can replace multiple versions of writing
   it in Julia (e.g. `x[:] == vec(x) == @cast ...`).
2) Use imports to make it more self-contained.

RomeoV committed:
Replace global imports with targeted ones.

Somehow this seems not so common in the Julia community, but if we want
to understand this file as a tutorial of some sorts I think it is very
helpful to understand why we are importing the packages.

RomeoV commented:
PS: Sorry for all the "outdated" tags - I pretty much took your suggestions exactly, but it seems when locally collecting the changes the github gui gets a bit confused. I'll look into it for next time.

mcabbott commented:
Thanks again for writing this up! Will add a link from the readme, too.
